File: apps/backend/analytics/insights/volume.py
"""Weekly volume trend insight.

Aggregates total volume (weight × reps_equivalent) per week and computes
linear slope and direction.
"""

from __future__ import annotations

from collections import defaultdict
from datetime import datetime

from queries import WorkoutRecord

# Assume each set contributes a fixed rep count when actual reps are absent.
_DEFAULT_REPS = 5


def compute(records: list[WorkoutRecord]) -> dict | None:
    """Return a volume_trend payload or None if insufficient data."""
    weekly: dict[str, float] = defaultdict(float)

    for r in records:
        if r.result != "success":
            continue
        reps = r.amrap_reps if r.amrap_reps and r.amrap_reps > 0 else _DEFAULT_REPS
        volume = r.weight * reps

        week_key = _week_key(r.recorded_at)
        if week_key:
            weekly[week_key] += volume

    if len(weekly) < 3:
        return None

    weeks = sorted(weekly.keys())
    volumes = [weekly[w] for w in weeks]

    slope = _linear_slope(volumes)
    direction = "up" if slope > 0.5 else ("down" if slope < -0.5 else "flat")

    return {
        "weeks": weeks,
        "volumes": volumes,
        "slope": round(slope, 2),
        "direction": direction,
    }


def _week_key(timestamp: str | None) -> str | None:
    if not timestamp:
        return None
    try:
        dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
        # ISO week: YYYY-Www
        return f"{dt.isocalendar().year}-W{dt.isocalendar().week:02d}"
    except ValueError:
        return None


def _linear_slope(values: list[float]) -> float:
    n = len(values)
    if n < 2:
        return 0.0
    xs = list(range(n))
    x_mean = sum(xs) / n
    y_mean = sum(values) / n
    num = sum((x - x_mean) * (y - y_mean) for x, y in zip(xs, values))
    den = sum((x - x_mean) ** 2 for x in xs)
    return num / den if den else 0.0

File: apps/backend/api/src/analytics/pipelines/volume.ts
/**
 * Weekly volume-trend insight (aggregate, exercise_id = null).
 *
 * Ports apps/backend/analytics/insights/volume.py. Aggregates total volume
 * (weight x reps_equivalent) per ISO week across all successful sets, then
 * reports the linear slope and a coarse up/down/flat direction.
 */

import type { WorkoutRecord } from '../record';
import { isoWeekKeyFromTimestamp } from '../iso-week';
import { pyRound } from './round';

/** Assume each set contributes a fixed rep count when actual reps are absent. */
const DEFAULT_REPS = 5;

export interface VolumeTrendPayload {
  readonly weeks: string[];
  readonly volumes: number[];
  readonly slope: number;
  readonly direction: 'up' | 'down' | 'flat';
}

export function computeVolume(records: readonly WorkoutRecord[]): VolumeTrendPayload | null {
  const weekly = new Map<string, number>();

  for (const r of records) {
    if (r.result !== 'success') continue;
    const reps = r.amrapReps && r.amrapReps > 0 ? r.amrapReps : DEFAULT_REPS;
    const volume = r.weight * reps;
    if (r.recordedAt === null) continue;
    const weekKey = isoWeekKeyFromTimestamp(r.recordedAt);
    if (weekKey === null) continue;
    weekly.set(weekKey, (weekly.get(weekKey) ?? 0) + volume);
  }

  if (weekly.size < 3) return null;

  const weeks = [...weekly.keys()].sort();
  const volumes = weeks.map((w) => weekly.get(w) ?? 0);

  const slope = linearSlope(volumes);
  const direction: VolumeTrendPayload['direction'] =
    slope > 0.5 ? 'up' : slope < -0.5 ? 'down' : 'flat';

  return {
    weeks,
    volumes,
    slope: pyRound(slope, 2),
    direction,
  };
}

/** Ordinary-least-squares slope of `values` against their indices 0..n-1,
 * matching volume.py:_linear_slope. */
function linearSlope(values: readonly number[]): number {
  const n = values.length;
  if (n < 2) return 0;

  let xMean = 0;
  let yMean = 0;
  for (let i = 0; i < n; i++) {
    xMean += i;
    yMean += values[i];
  }
  xMean /= n;
  yMean /= n;

  let num = 0;
  let den = 0;
  for (let i = 0; i < n; i++) {
    num += (i - xMean) * (values[i] - yMean);
    den += (i - xMean) ** 2;
  }
  return den !== 0 ? num / den : 0;
}
